src/services/invitation/index.ts

```diff
@@ -55,6 +119,109 @@ const plugin: FastifyPluginAsync = async (fastify, options) => {
+      // headerPairs: 2000             // Max number of header key=>value pairs (Default: 2000 - same as node's http).
+    },
+  });
+
+  fastify.route<{ Querystring: IdParam; Body: any }>({
+    method: HttpMethod.POST,
+    url: '/invitations/:id/upload_csv',
+    preHandler: fastify.verifyAuthentication,
+    handler: async (request, reply) => {
+      const {
+        member,
+        query: { id: parentId },
+        log,
+      } = request;
+      const errors: Error[] = [];
+      // check rights
+      try {
+        if (!parentId) {
+          throw new Error('User has passed an item located at its root filesystem');
+        }
+        //check permission
+        const repositories = buildRepositories();
+        const item = await repositories.itemRepository.get(parentId);
+        await validatePermission(repositories, PermissionLevel.Write, member, item);
+        const file = await request.file();
+
+        if (!file || file.mimetype != 'text/csv') {
+          throw new Error('An CSV file was not uploaded');
+        }
+        const { service: itemService } = items;
+        const stream = file.file;
+
+        const { rows, header } = await getCSV(stream);
+        const hasGrpCol = header.includes('group_name');
+        let groupNameToItem = new Map<string, Item>();
+        if (hasGrpCol) {
+          groupNameToItem = await rows.reduce(
+            async (result: Promise<Map<string, Item>>, rowInvite: CSVInvite) => {
+              const r = await result;
+              if (!r.has(rowInvite.group_name as string)) {
+                const newItem = await createNewFolder(member, rowInvite, parentId);
+                const itemCreated = await itemService.post(member, repositories, {
+                  item: newItem,
+                  parentId: parentId,
+                });
+                r.set(rowInvite.group_name as string, itemCreated);
+              }
+              return r;
+            },
+            Promise.resolve(new Map<string, Item>()),
+          );
+        }
+        const invitesPerGroup = await rows.reduce(
+          async (result: Promise<Map<string, Invitation[]>>, rowInvite: CSVInvite) => {
+            const r = await result;
+            let group_name = '';
+            if (hasGrpCol) group_name = rowInvite.group_name as string;
+            const invite = new Invitation();
+            invite.email = rowInvite['email'];
+            invite.permission = PermissionLevel.Read;
+            if (rowInvite.permission) invite.permission = rowInvite.permission;
+            if (!rowInvite.name) invite.name = '';
+            if (hasGrpCol) invite.item = groupNameToItem.get(group_name) as Item;
+            else invite.item = item;
+            if (!r.has(group_name)) {
+              r.set(group_name, [invite]);
+            } else {
+              r.get(group_name)?.push(invite);
+            }
+            return r;
+          },
+          Promise.resolve(new Map<string, Invitation[]>()),
```

Comment:
Why ?

Reply:
I have removed the async from this part, please let me know if this part still concerns you